win32ss/gdi/eng/pandisp.c

```diff
@@ -0,0 +1,484 @@
+
+    if (pandev->psoShadow)
+    {
+        EngUnlockSurface(pandev->psoShadow);
+        pandev->psoShadow = NULL;
+    }
+    if (pandev->hsurfShadow)
+    {
+        EngDeleteSurface(pandev->hsurfShadow);
+        pandev->hsurfShadow = NULL;
+    }
+    if (pandev->hsurf)
+    {
+        EngDeleteSurface(pandev->hsurf);
+        pandev->hsurf = NULL;
+    }
+    if (pandev->surfObjScreen)
+    {
+        EngUnlockSurface(pandev->surfObjScreen);
+        pandev->surfObjScreen = NULL;
+    }
+    if (pandev->enabledScreen)
+    {
+        GETPFN(DisableSurface)(pandev->dhpdevScreen);
+        pandev->enabledScreen = FALSE;
+    }
+}
+
+HSURF
+APIENTRY
+PanEnableSurface(
+    _In_ DHPDEV dhpdev)
+{
+    PPANDEV pandev = (PPANDEV)dhpdev;
+    HSURF surfScreen;
+
+    /* Move viewport to center of screen */
+    pandev->rclViewport.left = (pandev->szlDesktop.cx - pandev->szlScreen.cx) / 2;
+    pandev->rclViewport.top = (pandev->szlDesktop.cy - pandev->szlScreen.cy) / 2;
+    pandev->rclViewport.right = pandev->rclViewport.left + pandev->szlScreen.cx;
+    pandev->rclViewport.bottom  = pandev->rclViewport.top + pandev->szlScreen.cy;
+
+    /* Enable underlying device */
+    surfScreen = GETPFN(EnableSurface)(pandev->dhpdevScreen);
+    if (!surfScreen)
+    {
+        DPRINT1("Failed to enable underlying surface\n");
+        goto cleanup;
+    }
+    pandev->enabledScreen = TRUE;
+
+    pandev->surfObjScreen = EngLockSurface(surfScreen);
+    if (!pandev->surfObjScreen)
+    {
+        DPRINT1("Failed to lock underlying surface\n");
+        goto cleanup;
+    }
+
+    /* Create device surface */
+    pandev->hsurf = EngCreateDeviceSurface(NULL, pandev->szlDesktop, pandev->iBitmapFormat);
+    if (!pandev->hsurf)
+    {
+        DPRINT1("EngCreateDeviceSurface() failed\n");
+        goto cleanup;
+    }
+
+    if (!EngAssociateSurface(pandev->hsurf, pandev->hdev, HOOK_ALPHABLEND |
+                                                          HOOK_BITBLT |
+                                                          HOOK_COPYBITS |
+                                                          HOOK_GRADIENTFILL |
+                                                          HOOK_STROKEPATH |
+                                                          HOOK_SYNCHRONIZE |
+                                                          HOOK_TEXTOUT))
+
+    {
+        DPRINT1("EngAssociateSurface() failed\n");
+        goto cleanup;
+    }
+
+    /* Create shadow surface */
+    pandev->hsurfShadow = (HSURF)EngCreateBitmap(pandev->szlDesktop, pandev->szlDesktop.cx, pandev->iBitmapFormat, BMF_TOPDOWN, NULL);
+    if (!pandev->hsurfShadow)
+    {
+        DPRINT1("EngCreateBitmap() failed\n");
+        goto cleanup;
+    }
+
+    pandev->psoShadow = EngLockSurface(pandev->hsurfShadow);
+    if (!pandev->psoShadow)
+    {
+        DPRINT1("EngLockSurface() failed\n");
+        goto cleanup;
+    }
+
+    return pandev->hsurf;
+
+cleanup:
```

Comment:
More like `failure:` ? (using "cleanup" as a label sounds like a usual path taken even during success, as part of cleanup before returning)